app/controllers/application_controller.rb

```diff
@@ -1,12 +1,33 @@
 class ApplicationController < ActionController::Base
-  before_action :authenticate_user!
+  include Pundit
+  after_action :verify_authorized, unless: :devise_controller?
+  after_action :verify_policy_scoped, only: :index, unless: :devise_controller?
 
+  before_action :authenticate_user!
   before_action :configure_permitted_parameters, if: :devise_controller?
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
+  private
+
+  def user_not_authorized
+    flash[:alert] = "You are not authorized to perform this action."
+    redirect_back fallback_location: root_url
+  end
 
   protected
 
   def configure_permitted_parameters
     devise_parameter_sanitizer.permit(:sign_up, keys: [:username, :private, :name, :bio, :website, :avatar_image])
     devise_parameter_sanitizer.permit(:account_update, keys: [:username, :private, :name, :bio, :website, :avatar_image])
   end
+
+  def authorize_profile_access
+    unless policy(@user).profile_public?
+      flash.now[:alert] = "This user is private."
+      respond_to do |format|
+        format.html { render "show" }
+        format.json { head :no_content }
+      end
+    end
+  end
```

Comment:
This doesn't handle the cases where a user's profile is public but the logged user is following them.

Also, not sure if this is needed here since it should be handled by policies in the User Policy already:
- you call this method for User#show only
- if you delete this, since you already call `authorize @user` in User@show, you can set the policy for user#show to any rule you want. With these you will fall back on your feet and have the same implementation as what you are trying to do here.